Custom vaults docs (#5701)
Kong/developer.konghq.com: app/_how-tos/gateway/create-custom-vault.md

  - q: |
      {% include /gateway/vaults-format-faq.md type='question' %}
    a: |
      {% include /gateway/vaults-format-faq.md type='answer' %}

---

## Create the vault module

A custom vault backend is a Lua module with two required functions:

| Function | Description |
|----------|-------------|
| `init(conf)` | Called once at startup. Use it to validate config or open persistent connections. |
| `get(conf, resource, version)` | Called to resolve a secret reference. Returns the secret value as a string, or `nil` if not found. `resource` is the key portion of the `{vault://prefix/key}` reference. |

Create the directory structure:

```bash
mkdir -p kong/vaults/http
```

Create `kong/vaults/http.lua`:

```lua
echo '
local http = require("resty.http")
local cjson = require("cjson")

local function init()
end

local function get(conf, resource, version)
  local base_url = conf.base_url or "http://localhost:9876/sekretz"

  local httpc = http.new()
  local res, err = httpc:request_uri(base_url .. "/" .. resource, {
    method = "GET",
    headers = {
      ["Accept"] = "application/json",
    },
  })

  if not res then

>     ngx.log(ngx.WARN, "http vault: request failed for '", resource, "': ", err)

Author: Great catch

        ```
      icon_url: /assets/icons/gateway.svg

related_resources:
  - text: Secrets management
    url: /gateway/secrets-management/
  - text: Vault entity
    url: /gateway/entities/vault/
  - text: What can be stored as a secret?
    url: /gateway/entities/vault/#what-can-be-stored-as-a-secret

next_steps:
  - text: Vault entity reference
    url: /gateway/entities/vault/
  - text: Secrets management overview
    url: /gateway/secrets-management/

faqs:
  - q: When is `get` called?
    a: |
      {{site.base_gateway}} calls `get` whenever it needs to resolve a `{vault://...}` reference. With the default TTL, the resolved value is cached and `get` is not called again until the TTL expires. Set `KONG_VAULT_<NAME>_TTL=0` to disable caching and call `get` on every request.
  - q: Can my vault return a value other than a string?
    a: |
      No. `get` must return a string or `nil`. {{site.base_gateway}} stores the resolved value as a string and injects it into the plugin config field as-is.
  - q: |
      {% include /gateway/vaults-format-faq.md type='question' %}
    a: |
      {% include /gateway/vaults-format-faq.md type='answer' %}

---

## Create the vault module

> A custom vault backend is a Lua module with two required functions:

Reviewer: This needs a sentence or two here that explains why you'd want to use this custom vault over the other options.

Reviewer: ```suggestion
You can build a custom vault backend when your secrets live in a store that isn't one of the provided  [{{site.base_gateway}} Vaults](/gateway/entities/vault/#supported-vault-backends). 
For example, if your secrets live in an internal secrets service, a proprietary HTTP-based secret manager, or a legacy system that exposes secrets over an API only your organization uses.

A custom vault backend is a Lua module with two required functions:
```

         KONG_PG_DATABASE: kong
         KONG_LICENSE_DATA: ${KONG_LICENSE_DATA}
       depends_on:
         postgres:
           condition: service_healthy

     kong:
       image: kong/kong-gateway:latest
       environment:
         KONG_DATABASE: postgres
         KONG_PG_HOST: postgres
         KONG_PG_USER: kong
         KONG_PG_PASSWORD: kong
         KONG_PG_DATABASE: kong
         KONG_LICENSE_DATA: ${KONG_LICENSE_DATA}
         KONG_PROXY_LISTEN: "0.0.0.0:8000"
         KONG_ADMIN_LISTEN: "0.0.0.0:8001"
         KONG_VAULTS: bundled,http
         KONG_LUA_PACKAGE_PATH: /custom-code/?.lua;;
       ports:
         - "8000:8000"
         - "8001:8001"
       volumes:
         - ./kong:/custom-code/kong
       depends_on:
         postgres:
           condition: service_healthy
         kong-migrations:
           condition: service_completed_successfully
         secret-store:
           condition: service_healthy
       healthcheck:
         test: ["CMD", "kong", "health"]
         interval: 10s
         timeout: 5s
         retries: 10
       restart: unless-stopped' > docker-compose.yml
   ```

   The two key environment variables for custom vaults are:

>    - `KONG_VAULTS: bundled,http` — registers the built-in vaults plus the custom `http` vault

Reviewer: ```suggestion
   - `KONG_VAULTS: bundled,http`: Registers the built-in Vaults plus the custom `http` vault
```

>    - `KONG_LUA_PACKAGE_PATH: /custom-code/?.lua;;` — tells Kong where to find the Lua modules

Reviewer: ```suggestion
   - `KONG_LUA_PACKAGE_PATH: /custom-code/?.lua;;`: Tells {{site.base_gateway}} where to find the Lua modules
```

3. Start all services:

   ```bash
   docker compose up -d
   ```

## Create the Vault entity

> Create a Vault entity that configures the `http` vault backend. The `prefix` value (`http-vault`) is used in secret references:

Reviewer: ```suggestion
Create a [Vault entity](/gateway/entities/vault/) that configures the `http` vault backend. The `prefix` value (`http-vault`) is used in secret references:
```

{% entity_examples %}
entities:
  vaults:
    - name: http
      prefix: http-vault
      description: Custom HTTP vault backend
      config:
        base_url: http://secret-store:9876/sekretz
{% endentity_examples %}

> The `base_url` uses the Docker Compose service name (`secret-store`) as the hostname, since Kong and the secret store run in the same Docker network.

Reviewer: ```suggestion
The `base_url` uses the Docker Compose service name (`secret-store`) as the hostname, since {{site.base_gateway}} and the secret store run in the same Docker network.
```

## Configure the plugin with a vault reference

> Create a service and route, then enable the [Request Transformer Advanced](/plugins/request-transformer-advanced/) plugin. The `add.headers` list accepts items in `header-name:value` format. Here the vault reference is used as the entire item — Kong resolves `{vault://http-vault/x-from-vault}` to `X-From-Vault:top-secret-value`, which the plugin then parses into header name `X-From-Vault` and value `top-secret-value`.

Reviewer: ```suggestion
Create a Gateway Service and Route, then enable the [Request Transformer Advanced](/plugins/request-transformer-advanced/) plugin. The `add.headers` list accepts items in `header-name:value` format. Here the vault reference is used as the entire item; {{site.base_gateway}} resolves `{vault://http-vault/x-from-vault}` to `X-From-Vault:top-secret-value`, which the plugin then parses into header name `X-From-Vault` and value `top-secret-value`.
```

{% entity_examples %}
entities:
  services:
    - name: httpbin
      url: http://httpbin.konghq.com/anything
      routes:
        - name: httpbin-route
          paths:
            - /
          protocols:
            - http
      plugins:
        - name: request-transformer-advanced
          config:
            add:
              headers:
                - "{vault://http-vault/x-from-vault}"
{% endentity_examples %}

## Validate

Send a request through the proxy:

<!--vale off-->
{% validation request-check %}
url: /
status_code: 200
{% endvalidation %}
<!--vale on-->

> The [httpbin](https://httpbin.konghq.com) upstream echoes the full incoming request as a JSON response body. Check the `headers` object for `X-From-Vault`:

Reviewer: ```suggestion
The upstream echoes the full incoming request as a JSON response body. Check the `headers` object for `X-From-Vault`:
```
I thought we were removing links/mentions of httpbin a few years ago?
